source/fuzz/transformation_merge_function_returns.cpp

```diff
@@ -0,0 +1,646 @@
+  // predecessors are blocks in the loop (not further nested inside loops),
+  // which either return or are merge blocks of nested loops containing return
+  // instructions.
+  std::map<uint32_t, std::map<uint32_t, std::pair<uint32_t, uint32_t>>>
+      merge_blocks_to_returning_predecessors;
+
+  // Initialise the map, mapping each relevant merge block to an empty map.
+  for (uint32_t ret_block_id : return_blocks) {
+    uint32_t merge_block_id =
+        ir_context->GetStructuredCFGAnalysis()->LoopMergeBlock(ret_block_id);
+
+    while (merge_block_id != 0 &&
+           !merge_blocks_to_returning_predecessors.count(merge_block_id)) {
+      merge_blocks_to_returning_predecessors.emplace(
+          merge_block_id, std::map<uint32_t, std::pair<uint32_t, uint32_t>>());
+      merge_block_id = ir_context->GetStructuredCFGAnalysis()->LoopMergeBlock(
+          merge_block_id);
+    }
+  }
+
+  // Keep a map from all the new predecessors of the merge block of the new
+  // outer loop, to the related return value ids.
+  std::map<uint32_t, uint32_t> outer_merge_predecessors;
+
+  // Adjust the return blocks and add the related information to the map or
+  // |outer_merge_predecessors| set.
+  for (uint32_t ret_block_id : return_blocks) {
+    auto ret_block = ir_context->get_instr_block(ret_block_id);
+
+    // Get the return value id (if the function is not void).
+    uint32_t ret_val_id =
+        function_type->AsVoid()
+            ? 0
+            : ret_block->terminator()->GetSingleWordInOperand(0);
+
+    uint32_t merge_block_id =
+        ir_context->GetStructuredCFGAnalysis()->LoopMergeBlock(ret_block_id);
+
+    // Add a new entry to the map corresponding to the merge block of the
+    // innermost enclosing loop (or that of the new outer loop if there is no
+    // enclosing loop).
+    if (merge_block_id != 0) {
+      merge_blocks_to_returning_predecessors[merge_block_id].emplace(
+          ret_block_id,
+          std::pair<uint32_t, uint32_t>(ret_val_id, constant_true));
+    } else {
+      // If there is no enclosing loop, the block will branch to the merge block
+      // of the new outer loop.
+      merge_block_id = message_.outer_return_id();
+      outer_merge_predecessors.emplace(ret_block_id, ret_val_id);
+    }
+
+    // Replace the return instruction with an unconditional branch.
+    ret_block->terminator()->SetOpcode(SpvOpBranch);
+    ret_block->terminator()->SetInOperands(
+        {{SPV_OPERAND_TYPE_RESULT_ID, {merge_block_id}}});
+  }
+
+  // Get a list of all the relevant merge blocks.
+  std::vector<uint32_t> merge_blocks;
+  merge_blocks.reserve(merge_blocks_to_returning_predecessors.size());
+  for (const auto& entry : merge_blocks_to_returning_predecessors) {
+    merge_blocks.emplace_back(entry.first);
+  }
+  // Sort the list so that deeper merge blocks come first.
+  std::sort(merge_blocks.begin(), merge_blocks.end(),
+            ComparatorDeepBlocksFirst(ir_context));
```

Comment:
Presumably this is required for correctness; can you very briefly comment on why?

Reply:
Done

Reply:
(the reason is that, when adjusting a merge block, all the information about return blocks or other merge blocks from which the function might be returning need to be present in the map, so we need to consider the deepest ones first)